File: compiler_python/CodeGenerator.java
// package CourseWord;

import java.util.List;
import java.util.ArrayList;
import java.util.Map;

public class CodeGenerator {
    private StringBuilder code;

    private List<Node> nodes;
    private Map<String, Type> variablesTable;

    public CodeGenerator(List<Node> nodes, Map<String, Type> variablesTable) {
        this.code = new StringBuilder();
        this.nodes = nodes;
        this.variablesTable = variablesTable;
    }

    public String generate() {
        code.append(".data\n");
        for (Map.Entry<String, Type> entry : variablesTable.entrySet()) {
            final Type variableType = entry.getValue();
            if (!(variableType == Type.BOOLEAN || variableType == Type.NUM)) {
                continue;
            }
            final String variableName = entry.getKey();
            code.append("\t")
                .append(variableName)
                .append(" ")
                .append(variableType == Type.BOOLEAN ? "db" : "dd")
                .append(" 0")
                .append("\n");
        }
        code.append("\n");

        code.append(".code\n");

        List<Node> intoMain = new ArrayList<>();
        boolean inFunction = false;
        for (Node node : nodes) {
            switch (node.name) {
                case FUNCTION_NEW:
                    inFunction = true;

                    final FunctionNew nodeFuncNew = (FunctionNew) node;
                    code.append(nodeFuncNew.name)
                        .append(" proc\n");
                    code.append("\tpush ebp").append("\n");
                    code.append("\tmov ebp, esp").append("\n");
                    code.append("\n");
                    for (int i = 0; i < nodeFuncNew.listOfArguments.size(); i++) {
                        final String name = nodeFuncNew.listOfArguments.get(i).name;
                        final int stackIndex =
                            8 + (nodeFuncNew.listOfArguments.size() - i - 1) * 4;
                        code.append("\tmov eax, dword ptr[ebp + ")
                            .append(stackIndex)
                            .append("]\n");
                        code.append("\tmov ")
                            .append(name)
                            .append(", eax\n");
                    }
                    code.append("\n");
                    break;
                case FUNCTION_END:
                    inFunction = false;

                    final FunctionEnd nodeFuncEnd = (FunctionEnd) node;
                    code.append("\n");
                    code.append(nodeFuncEnd.name + "End:\n");
                    code.append("\tmov esp, ebp").append("\n")
                        .append("\tpop ebp").append("\n")
                        .append("\tret ").append((4*nodeFuncEnd.amountOfArguments)).append("\n");
                    code.append(nodeFuncEnd.name)
                        .append(" endp\n");
                    break;
                case STATEMENT:
                    if (inFunction) {
                        code.append(((Statement)node).code);
                    } else {
                        intoMain.add(node);
                    }
                    break;
            }
        }
        code.append("\n");

        code.append("\n");
        code.append("main proc").append("\n");
        code.append("\tpush ebp").append("\n");
        code.append("\tmov ebp, esp").append("\n");
        code.append("\n");
        for (Node node : intoMain) {
            code.append(((Statement)node).code);
        }
        code.append("\tmov esp, ebp").append("\n");
        code.append("\tpop ebp").append("\n");
        code.append("\tret").append("\n");
        code.append("main endp").append("\n");

        return code.toString();
    }
}

File: compiler_python/CodeGenerator.cpp
#include "CodeGenerator.h"


CodeGenerator::CodeGenerator(
        const std::vector<Node*> nodes,
        const std::map<std::string, Type> variablesTable) :
    m_Nodes(nodes), m_VariablesTable(variablesTable) {}


std::string CodeGenerator::generate() const {
    std::stringstream code;
    code << ".386\n"
         << ".model flat\n"
         << ".stack 4096\n"
         << "include masm32\\include\\kernel.inc\n"
         << "includelib masm32\\lib\\kernel.lib\n";

    code << ".data\n";
    for (const auto& variable : m_VariablesTable) {
        const Type variableType = variable.second;
        if (!(variableType == Type_Bool || variableType == Type_Number)) {
            continue;
        }

        const std::string variableName = variable.first;
        code << "\t"
             << variableName
             << " "
             << (variableType == Type_Bool ? "db" : "dd")
             << " 0"
             << "\n";
    }
    code << "\n";

    code << ".code\n";
    std::vector<Node*> intoMain;
    bool inFunction = false;
    for (Node* node : m_Nodes) {
        switch (node->type) {
            case FunctionNew: {
                inFunction = true;

                const NodeFunctionNew* const nodeFunctionNew =
                    static_cast<const NodeFunctionNew* const>(node);
                code << nodeFunctionNew->name
                     << " proc\n"
                     << "\tpush ebp\n"
                     << "\tmov ebp, esp\n"
                     << "\n";
                for (unsigned int i = 0; i < nodeFunctionNew->arguments.size(); i++) {
                    const std::string name = nodeFunctionNew->arguments.at(i).name;
                    const Argument arg = nodeFunctionNew->arguments.at(i);
                    const unsigned int stackIndex =
                        8 + (nodeFunctionNew->arguments.size() - i - 1) * 4;
                    code << "\tmov eax, dword prt[ebp + "
                         << stackIndex
                         << "]\n";
                    code << "\tmov "
                         << name
                         << ", eax\n";
                }
                code << "\n";
                break;
            }
            case FunctionEnd: {
                inFunction = false;

                const NodeFunctionEnd* const nodeFunctionEnd =
                    static_cast<const NodeFunctionEnd* const>(node);
                code << "\n"
                     << nodeFunctionEnd->name << "End:\n"
                     << "\tmov esp, ebp\n"
                     << "\tpop ebp\n"
                     << "\tret "
                     << (4 * nodeFunctionEnd->amountOfArguments)
                     << "\n"
                     << nodeFunctionEnd->name
                     << " endp\n";
                break;
            }
            case Statement:
                if (inFunction) {
                    const NodeStatement* const nodeStatement =
                        static_cast<const NodeStatement* const>(node);
                    code << nodeStatement->code;
                } else {
                    intoMain.push_back(node);
                }
        }
    }
    code << "\n";

    code << "\n"
         << "MAIN:\n"
         << "main proc:\n"
         << "\tpush ebp\n"
         << "\tmov ebp, esp\n"
         << "\n";
    for (const Node* node : intoMain) {
        const NodeStatement* const nodeStatement =
            static_cast<const NodeStatement* const>(node);
        code << nodeStatement->code;
    }
    code << "\tmov esp, ebp\n"
         << "\tpop ebp\n"
         << "\tret\n"
         << "main endp\n";

    return code.str();
}
